editors/sc-ide/core/settings/serialization.cpp

```diff
@@ -63,18 +63,36 @@ static QVariant parseTextFormat( const YAML::Node & node )
     std::string val;
     QTextCharFormat fm;
 
+    // Some really old themes have missing color: or background: entries, which
+    // caused problems here (bug #4218). We default to black if text color is
+    // not specified.
+    fm.setForeground(Qt::black);
+
     const Node ncolor = node[ "color" ];
     if(ncolor && ncolor.IsScalar()) {
         val = ncolor.as<std::string>();
         fm.setForeground(QColor(val.c_str()));
     }
 
     const Node nbg = node[ "background" ];
+    bool have_background = false;
     if(nbg && nbg.IsScalar()) {
         val = nbg.as<std::string>();
         QColor color(val.c_str());
-        if(color.isValid())
+        if(color.isValid()) {
+            have_background = true;
             fm.setBackground(color);
+        }
+    }
+
+    // If background color is not specified, pick either black or white
+    // depending on what contrasts with the foreground color.
+    if (!have_background) {
+        if (fm.foreground().color().value() < 127) {
+            fm.setBackground(Qt::white);
+        } else {
+            fm.setBackground(Qt::black);
+        }
```

Comment:
this logic could be clearer (the `bool` local variable is a code smell). there are a couple ways you could do that, this is one.

While we're looking at this -- why is the `isValid` check used for the background color but not the foreground color? You could pull out this unwrapping logic into a separate function `QColor tryGetColor(node, "color")`, returning a default-constructed `QColor` on failure (which is invalid by default). Then this whole section can be written as:

```cpp
auto fgColor = tryGetColor(node, "color");
fm.setForeground(fgColor.isValid() ? fgColor : Qt::black);

auto bgColor = tryGetColor(node, "background");
if (bgColor.isValid()) {
    fm.setBackground(bgColor);
} else if (fm.foreground().color().value < 127)) {
    fm.setBackground(Qt::white);
} else {
    fm.setBackground(Qt::black);
}
```

would also think about adding a log line for the isn't-valid case if there's a pattern for that nearby